src/librustc/ty/context.rs

```diff
@@ -1512,6 +1512,24 @@ impl<'tcx> TyCtxt<'tcx> {
                 .subst(*self, self.mk_substs([self.lifetimes.re_static.into()].iter())),
         )
     }
+
+    /// Returns a displayable description and article for the given `def_id` (e.g. `("a", "struct")`).
+    pub fn article_and_description(&self, def_id: DefId) -> (&'static str, &'static str) {
+        match self.def_key(def_id).disambiguated_data.data {
+            DefPathData::TypeNs(..) | DefPathData::ValueNs(..) | DefPathData::MacroNs(..) => {
+                let kind = self.def_kind(def_id).unwrap();
+                (kind.article(), kind.descr(def_id))
```

Comment:
I wish I saw this earlier - I think you should *always* call `def_kind` and only look at the `def_key` or other things if `def_kind` returned `None`.

The reason is that `DefKind` should eventually encompass every possible `DefId`, and the current situation is a transitional one (since `DefKind` was split out of name resolution, which only needs the current set of possible `DefKind`s).

Reply:
I don't know if there is an issue for finishing `DefKind`, but feel free to.

Ideally `def_key` would be used in as few places as possible, as it's really the "name" of a definition more than anything else.